File: src/main/java/com/dude/dms/ui/components/tags/TagSelector.java
package com.dude.dms.ui.components.tags;

import com.dude.dms.backend.data.Tag;
import com.dude.dms.backend.service.TagService;
import com.vaadin.flow.component.checkbox.Checkbox;
import com.vaadin.flow.component.grid.Grid;

import java.util.List;
import java.util.Set;

public class TagSelector extends Grid<Tag> {

    public TagSelector(TagService tagService) {
        List<Tag> tags = tagService.findAll();

        setItems(tags);
        addColumn(Tag::getName).setHeader("Tags").setAutoWidth(true);
        setSizeFull();

        setSelectionMode(SelectionMode.MULTI);
        addItemClickListener(event -> {
            if (asMultiSelect().isSelected(event.getItem())) {
                deselect(event.getItem());
            } else {
                select(event.getItem());
            }
        });
    }

    public void setSelectedTags(Iterable<Tag> tags) {
        asMultiSelect().select(tags);
    }

    public void setContainedTags(String rawText) {
        addComponentColumn(tag -> {
            Checkbox checkbox = new Checkbox(rawText != null && rawText.contains(tag.getName()));
            checkbox.setReadOnly(true);
            return checkbox;
        }).setHeader("Contained").setAutoWidth(true);
    }

    public Set<Tag> getSelectedTags() {
        return asMultiSelect().getSelectedItems();
    }
}

File: src/main/java/com/dude/dms/ui/components/tags/TagSelector.kt
package com.dude.dms.ui.components.tags

import com.dude.dms.backend.containers.TagContainer
import com.dude.dms.backend.service.TagService
import com.dude.dms.brain.t
import com.vaadin.flow.component.checkbox.Checkbox
import com.vaadin.flow.component.grid.Grid
import com.vaadin.flow.component.html.Label
import dev.mett.vaadin.tooltip.Tooltips

class TagSelector(private val tagService: TagService) : Grid<TagContainer>() {

    var selectedTags: Set<TagContainer>
        get() = asMultiSelect().selectedItems
        set(tags) {
            setItems(tags.sortedBy { it.tag.name }.toSet().plus(tagService.findAll().map { TagContainer(it) }.sortedBy { it.tag.name }.toSet()))
            asMultiSelect().deselectAll()
            asMultiSelect().select(tags)
        }

    var rawText: String? = null

    init {
        setSelectionMode(SelectionMode.MULTI)
        if (selectedTags.isNullOrEmpty()) {
            setItems(tagService.findAll().map { TagContainer(it) }.sortedBy { it.tag.name })
        }
        val tooltip = Tooltips.getCurrent()
        addComponentColumn { tagContainer ->
            Label(tagContainer.tag.name).apply { tagContainer.tagOrigin?.let { tooltip.setTooltip(this, it) } }
        }.setHeader(t("tags")).setKey("tag").isAutoWidth = true
        addComponentColumn {
            Checkbox(rawText?.contains(it.tag.name) ?: false).apply { isReadOnly = true }
        }.setHeader(t("contained")).setAutoWidth(true).setKey("contained").isVisible = false
        setSizeFull()
        addItemClickListener { event ->
            if (asMultiSelect().isSelected(event.item)) {
                deselect(event.item)
            } else {
                select(event.item)
            }
        }
    }

    fun showContainedTags(show: Boolean) {
        getColumnByKey("contained").isVisible = show
    }
}